#!/bin/sh /usr/share/dpatch/dpatch-run
## 09-initscript.dpatch by Michael Meskes <meskes@debian.org>
##
## DP: Several initscript changes.

@DPATCH@

diff -Naurp virtualbox-ose-2.2.4-dfsg.orig/src/VBox/Installer/linux/vboxdrv.sh.in virtualbox-ose-2.2.4-dfsg/src/VBox/Installer/linux/vboxdrv.sh.in
--- virtualbox-ose-2.2.4-dfsg.orig/src/VBox/Installer/linux/vboxdrv.sh.in	2009-06-03 09:24:58.000000000 +0200
+++ virtualbox-ose-2.2.4-dfsg/src/VBox/Installer/linux/vboxdrv.sh.in		2009-06-03 09:29:45.000000000 +0200
@@ -19,8 +19,8 @@
 #
 ### BEGIN INIT INFO
 # Provides:       vboxdrv
-# Required-Start: $syslog
-# Required-Stop:
+# Required-Start: $remote_fs $network
+# Required-Stop:  $remote_fs
 # Default-Start:  2 3 4 5
 # Default-Stop:   0 1 6
 # Short-Description: VirtualBox Linux kernel module
@@ -32,7 +32,9 @@
 LOG="/var/log/vbox-install.log"
 NOLSB=%NOLSB%
 
-[ -f /lib/lsb/init-functions ] || NOLSB=yes
+# Include virtualbox-ose defaults if available
+[ -f /etc/default/virtualbox-ose ] && . /etc/default/virtualbox-ose
+[ -f /lib/lsb/init-functions ] && NOLSB=
 [ -f /etc/vbox/vbox.cfg ] && . /etc/vbox/vbox.cfg
 
 if [ -n "$INSTALL_DIR" ]; then
@@ -64,11 +66,14 @@
         log_failure_msg "$1"
     }
     succ_msg() {
-        log_success_msg " done."
+	log_end_msg 0
     }
     begin_msg() {
         log_daemon_msg "$@"
     }
+    prog_msg() {
+	log_progress_msg "$1"
+    }
 else
     if [ "$system" = "redhat" ]; then
         . /etc/init.d/functions
@@ -146,15 +151,26 @@
 
 start()
 {
-    begin_msg "Starting VirtualBox kernel module"
+    begin_msg "Starting VirtualBox kernel modules"
     if ! running vboxdrv; then
+      if [ "$LOAD_VBOXDRV_MODULE" = 1 ]; then
         if ! rm -f $DEVICE; then
             failure "Cannot remove $DEVICE"
         fi
         if ! modprobe vboxdrv > /dev/null 2>&1; then
+	  if ! find /lib/modules/`uname -r` -name "vboxdrv\.*" 2>/dev/null|grep -q vboxdrv; then
+	    failure "No suitable module for running kernel found"
+	  else
             failure "modprobe vboxdrv failed. Please use 'dmesg' to find out why"
-        fi
-        sleep .2
+	  fi
+	else
+	  prog_msg vboxdrv
+        fi
+      else
+	succ_msg
+	return
+      fi
+      sleep .2
     fi
     # ensure the character special exists
     if [ ! -c $DEVICE ]; then
@@ -182,23 +198,35 @@
         rmmod vboxdrv 2>/dev/null
         failure "Cannot change owner $GROUPNAME for device $DEVICE"
     fi
-    if ! modprobe vboxnetflt > /dev/null 2>&1; then
-        failure "modprobe vboxnetflt failed. Please use 'dmesg' to find out why"
+    if [ "$LOAD_VBOXDRV_MODULE" = 1 ]; then
+	if ! modprobe vboxnetflt > /dev/null 2>&1; then
+		if ! find /lib/modules/`uname -r` -name "vboxnetflt\.*" 2>/dev/null|grep -q vboxnetflt; then
+		    failure "No suitable vboxnetflt module for running kernel found"
+		else
+        	    failure "modprobe vboxnetflt failed. Please use 'dmesg' to find out why"
+		fi
+	else
+		prog_msg vboxnetflt
+	fi
     fi
     succ_msg
 }
 
 stop()
 {
-    begin_msg "Stopping VirtualBox kernel module"
+    begin_msg "Stopping VirtualBox kernel modules"
     if running vboxdrv; then
         if running vboxnetflt; then
             if ! rmmod vboxnetflt 2>/dev/null; then
                 failure "Cannot unload module vboxnetflt"
+	    else
+		prog_msg vboxnetflt
             fi
         fi
         if ! rmmod vboxdrv 2>/dev/null; then
             failure "Cannot unload module vboxdrv"
+	else
+	    prog_msg vboxdrv
         fi
         if ! rm -f $DEVICE; then
             failure "Cannot unlink $DEVICE"
@@ -327,14 +355,11 @@
     stop
     start
     ;;
-setup)
-    setup
-    ;;
 status)
     dmnstatus
     ;;
 *)
-    echo "Usage: $0 {start|stop|stop_vms|restart|force-reload|status|setup}"
+    echo "Usage: $0 {start|stop|stop_vms|restart|force-reload|status}"
     exit 1
 esac
 
